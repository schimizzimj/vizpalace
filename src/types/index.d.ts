interface BarData {
  label: string;
  value: number;
}

interface LineData {
  date: Date;
  value: number;
}

interface ChartOptions {
  width: number;
  height: number;
  margin?: {
    top: number;
    right: number;
    bottom: number;
    left: number;
  };
  animationDuration?: number;
}
