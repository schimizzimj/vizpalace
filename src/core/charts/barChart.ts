import * as d3 from "d3";
import {
  initializeChartDimensions,
  createSVG,
  appendXAxis,
  appendYAxis,
} from "../base/chartUtils";
import { BarData, ChartOptions } from "../../types";

export function createBarChart(
  element: HTMLElement,
  data: BarData[],
  userOptions: Partial<ChartOptions> = {}
): void {
  // Merge default options with user options
  const options = initializeChartDimensions(userOptions);
  let { width, height, animationDuration } = options;
  const svg = createSVG(element, options);
  animationDuration = animationDuration ?? 1000;

  const xScale = d3
    .scaleBand()
    .domain(data.map((d) => d.label))
    .range([0, width])
    .padding(0.1);

  const yMax =
    d3.max(data, (d) =>
      d.value !== undefined && d.value !== null ? d.value : 0
    ) || 0;

  const yScale = d3.scaleLinear().domain([0, yMax]).range([height, 0]);

  const bars = svg
    .selectAll(".bar")
    .data(data)
    .enter()
    .append("rect")
    .attr("class", "bar")
    .attr("x", (d) => xScale(d.label) || 0)
    .attr("y", height)
    .attr("width", xScale.bandwidth())
    .attr("height", 0);

  bars
    .transition()
    .duration(animationDuration)
    .ease(d3.easeCubicInOut)
    .attr("y", (d) => yScale(d.value ?? 0))
    .attr("height", (d) => height - yScale(d.value ?? 0));

  appendXAxis(svg, xScale, height);
  appendYAxis(svg, yScale);
}
