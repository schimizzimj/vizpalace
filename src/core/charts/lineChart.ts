import * as d3 from "d3";
import {
  initializeChartDimensions,
  createSVG,
  appendXAxis,
  appendYAxis,
} from "../base/chartUtils";

export function createLineChart(
  element: HTMLElement,
  data: LineData[],
  userOptions: Partial<ChartOptions> = {}
): void {
  const options = initializeChartDimensions(userOptions);
  let { width, height, animationDuration } = options;

  const svg = createSVG(element, options);
  animationDuration = animationDuration ?? 0;

  const xScale = d3
    .scaleTime()
    .domain(d3.extent(data, (d) => d.date) as [Date, Date])
    .range([0, width]);

  const yMax =
    d3.max(data, (d) =>
      d.value !== undefined && d.value !== null ? d.value : 0
    ) || 0;

  const yScale = d3.scaleLinear().domain([0, yMax]).range([height, 0]);

  // Create the line generator
  const line = d3
    .line<LineData>()
    .x((d) => xScale(d.date))
    .y((d) => yScale(d.value ?? 0));

  // Draw the line
  const linePath = svg
    .append("path")
    .datum(data)
    .attr("class", "line")
    .attr("d", line)
    .attr("fill", "none")
    .attr("stroke", "steelblue")
    .attr("stroke-width", 1.5);

  if (linePath.node()?.getTotalLength) {
    const totalLength = linePath.node()?.getTotalLength() ?? 0;
    linePath
      .attr("stroke-dasharray", `${totalLength} ${totalLength}`)
      .attr("stroke-dashoffset", totalLength)
      .transition()
      .duration(animationDuration)
      .ease(d3.easeLinear)
      .attr("stroke-dashoffset", 0);
  }

  appendXAxis(svg, xScale, height);
  appendYAxis(svg, yScale);
}
