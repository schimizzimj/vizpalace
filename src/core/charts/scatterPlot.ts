import * as d3 from "d3";
import { ChartData, ScatterPlotOptions } from "../../types";
import { createSVG, initializeScatterPlotDimensions } from "../base/chartUtils";
import { appendXAxis, appendYAxis } from "../base/axes";
import { isScaleBand } from "../utils/utils";

export function createScatterPlot<T = number | string | Date>(
  element: HTMLElement,
  data: ChartData<T>,
  userOptions: Partial<ScatterPlotOptions> = {}
) {
  const options = initializeScatterPlotDimensions(userOptions);
  let { chartDimensions, animationDuration, xAxis, yAxis } = options;
  const { width, height } = chartDimensions;
  const svg = createSVG(element, options);
  const duration = animationDuration ?? 0;

  const allDataPoints = data.flatMap((series) => series.values);

  let xScale:
    | d3.ScaleBand<string>
    | d3.ScaleTime<number, number>
    | d3.ScaleLinear<number, number>;

  if (typeof allDataPoints[0].x === "string") {
    xScale = d3
      .scaleBand()
      .padding(0.1)
      .domain(allDataPoints.map((d) => d.x as string));
  } else if (allDataPoints[0].x instanceof Date) {
    xScale = d3
      .scaleTime()
      .domain(d3.extent(allDataPoints, (d) => d.x as Date) as [Date, Date]);
  } else {
    xScale = d3
      .scaleLinear()
      .domain(
        d3.extent(allDataPoints, (d) => d.x as number) as [number, number]
      );
  }

  xScale.range([0, width]);

  const yMin = d3.min(allDataPoints, (d) => d.y ?? 0) ?? 0;
  const yMax = d3.max(allDataPoints, (d) => d.y ?? 0) ?? 0;

  const yScale = d3.scaleLinear().domain([yMin, yMax]).range([height, 0]);

  let circlesWrapper: d3.Selection<SVGGElement, any, any, any> =
    svg.select(".circles-wrapper");

  if (circlesWrapper.empty()) {
    circlesWrapper = svg.append("g").attr("class", "circles-wrapper");
  }

  const circles = circlesWrapper
    .selectAll<SVGCircleElement, any>(".circle")
    .data(allDataPoints);

  circles.exit().remove();

  circles
    .enter()
    .append("circle")
    .attr("class", "circle")
    .attr("cx", (d) => {
      if (isScaleBand(xScale)) {
        return (xScale(d.x as string) ?? 0) + xScale.bandwidth() / 2;
      } else {
        return xScale(d.x as any) ?? 0;
      }
    })
    .attr("cy", (d) => yScale(d.y ?? 0) ?? 0)
    .attr("r", 5)
    .attr("fill", "steelblue")
    .merge(circles)
    .transition()
    .duration(duration)
    .attr("cx", (d) => {
      if (isScaleBand(xScale)) {
        return (xScale(d.x as string) ?? 0) + xScale.bandwidth() / 2;
      } else {
        return xScale(d.x as any) ?? 0;
      }
    })
    .attr("cy", (d) => yScale(d.y ?? 0) ?? 0);

  appendXAxis(svg, xScale, chartDimensions, xAxis ?? {});
  appendYAxis(svg, yScale, chartDimensions, yAxis ?? {});
}
