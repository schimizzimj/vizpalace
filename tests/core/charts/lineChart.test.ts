import { assert, expect, test, describe, beforeEach, afterEach } from "vitest";
import { createLineChart } from "../../../src/core/charts/lineChart";
import { ChartData } from "../../../src/types";

describe("lineChart", () => {
  let container: HTMLDivElement;
  let data: ChartData;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
  });

  describe("single series", () => {
    beforeEach(() => {
      data = [
        {
          name: "Series 1",
          values: [
            { x: new Date("2020-01-01"), y: 10 },
            { x: new Date("2020-01-02"), y: 20 },
            { x: new Date("2020-01-03"), y: 30 },
          ],
        },
      ];
    });

    test("it should render a line chart", () => {
      createLineChart(container, data, {});

      const svg = container.querySelector("svg");
      assert(svg);

      const lines = svg.querySelectorAll(".line");
      expect(lines.length).toBe(1);
    });

    test("it should use custom width and height", () => {
      const config = {
        width: 100,
        height: 100,
        margin: { top: 0, bottom: 0, left: 0, right: 0 },
      };
      createLineChart(container, data, config);

      const svg = container.querySelector("svg");
      assert(svg);
      expect(svg.getAttribute("width")).toBe("100");
      expect(svg.getAttribute("height")).toBe("100");
    });

    test("it should use custom margins", () => {
      const config = {
        width: 100,
        height: 100,
        margin: { top: 10, right: 10, bottom: 10, left: 10 },
      };
      createLineChart(container, data, config);

      const svg = container.querySelector("svg");
      assert(svg);
      expect(svg.getAttribute("width")).toBe(
        `${config.width + config.margin.left + config.margin.right}`
      );
      expect(svg.getAttribute("height")).toBe(
        `${config.height + config.margin.top + config.margin.bottom}`
      );
    });

    test("it should look visually correct", async () => {
      createLineChart(container, data, {});
      await new Promise((resolve) => setTimeout(resolve, 1000));
      const result = container.innerHTML;
      expect(result).toMatchSnapshot();
    });
  });

  describe("multiple series", () => {
    beforeEach(() => {
      data = [
        {
          name: "Series 1",
          values: [
            { x: new Date("2020-01-01"), y: 10 },
            { x: new Date("2020-01-02"), y: 20 },
            { x: new Date("2020-01-03"), y: 30 },
          ],
        },
        {
          name: "Series 2",
          values: [
            { x: new Date("2020-01-01"), y: 30 },
            { x: new Date("2020-01-02"), y: 20 },
            { x: new Date("2020-01-03"), y: 10 },
          ],
        },
      ];
    });

    test("it should render a line chart", () => {
      createLineChart(container, data, {});

      const svg = container.querySelector("svg");
      assert(svg);

      const lines = svg.querySelectorAll(".line");
      expect(lines.length).toBe(2);
    });

    test("it should look visually correct", async () => {
      createLineChart(container, data, {});
      await new Promise((resolve) => setTimeout(resolve, 1000));
      const result = container.innerHTML;
      expect(result).toMatchSnapshot();
    });
  });
});
