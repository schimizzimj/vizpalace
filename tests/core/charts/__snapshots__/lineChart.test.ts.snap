// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lineChart > multiple series > it should look visually correct 1`] = `"<svg width=\\"660\\" height=\\"450\\"><g class=\\"chart-wrapper\\" transform=\\"translate(40, 20)\\"><g class=\\"lines-wrapper\\"><path class=\\"line\\" fill=\\"none\\" stroke=\\"black\\" d=\\"M0,400L300,200L600,0\\" stroke-dasharray=\\"none\\" stroke-dashoffset=\\"0\\" stroke-width=\\"1.5\\"></path><path class=\\"line\\" fill=\\"none\\" stroke=\\"black\\" d=\\"M0,0L300,200L600,400\\" stroke-dasharray=\\"none\\" stroke-dashoffset=\\"0\\" stroke-width=\\"1.5\\"></path></g><g class=\\"x-axis\\" transform=\\"translate(0, 400)\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"middle\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M0.5,6V0.5H600.5V6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(25.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">06 PM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(100.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">2020</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(175.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">06 AM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(250.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">12 PM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(325.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">06 PM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(400.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">Thu 02</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(475.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">06 AM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(550.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">12 PM</text></g></g><g class=\\"y-axis\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"end\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M-6,400.5H0.5V0.5H-6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,400.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">10</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,360.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">12</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,320.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">14</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,280.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">16</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,240.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">18</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,200.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">20</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,160.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">22</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,120.50000000000001)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">24</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,80.49999999999999)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">26</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,40.49999999999999)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">28</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,0.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">30</text></g></g></g></svg>"`;

exports[`lineChart > single series > it should look visually correct 1`] = `"<svg width=\\"660\\" height=\\"450\\"><g class=\\"chart-wrapper\\" transform=\\"translate(40, 20)\\"><g class=\\"lines-wrapper\\"><path class=\\"line\\" fill=\\"none\\" stroke=\\"black\\" d=\\"M0,400L300,200L600,0\\" stroke-dasharray=\\"none\\" stroke-dashoffset=\\"0\\" stroke-width=\\"1.5\\"></path></g><g class=\\"x-axis\\" transform=\\"translate(0, 400)\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"middle\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M0.5,6V0.5H600.5V6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(25.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">06 PM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(100.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">2020</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(175.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">06 AM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(250.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">12 PM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(325.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">06 PM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(400.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">Thu 02</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(475.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">06 AM</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(550.5,0)\\"><line stroke=\\"currentColor\\" y2=\\"6\\"></line><text fill=\\"currentColor\\" y=\\"9\\" dy=\\"0.71em\\">12 PM</text></g></g><g class=\\"y-axis\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"end\\"><path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M-6,400.5H0.5V0.5H-6\\"></path><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,400.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">10</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,360.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">12</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,320.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">14</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,280.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">16</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,240.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">18</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,200.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">20</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,160.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">22</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,120.50000000000001)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">24</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,80.49999999999999)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">26</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,40.49999999999999)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">28</text></g><g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,0.5)\\"><line stroke=\\"currentColor\\" x2=\\"-6\\"></line><text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">30</text></g></g></g></svg>"`;
